### PR TITLE
Add undefined resp callback

### PR DESCRIPTION
**Pull request by chinglee-iot**
This PR was created to support cellular modem porting with modem specific response in #45.

* Add CELLULAR_MODEM_NO_EPS_NETWORK for cellular modem without ESP network. 
Developer can defined the macro in cellular_config.h to disable the EPS network registration query in Cellular_CommonGetServiceStatus.
* Add _Cellular_RegisterUndefinedRespCallback API in cellular_common.c for cellular modem to handle undefined response.
The socket connected URC code for Quectel GSM modem is "\<index\>, CONNECT OK"
The support URC code format is either fixed string or string with prefix. This function pass the undefined response to cellular modem porting.
* Add _Cellular_AtcmdRequestSuccessToken API in cellular_pkthandler.c for cellular modem with different success Token.
The socket close AT command response to indicate command success for Quectel GSM modem  is "\<index\>, CLOSE OK". 
It is not a fixed string. This API makes use of a temporary success token table to indicate command success.

### CODE REVIEW

**Comment by ActoryOu** [comment]
Hi,
Why/Where should users define this option CELLULAR_MODEM_NO_EPS_NETWORK?
Could you add an example/explaination in configuration file for future use?

**Review by ActoryOu — DISMISSED** [review]

File: source/include/common/cellular_common.h [hunk 1]
```diff
@@ -208,6 +208,19 @@ typedef CellularPktStatus_t ( * CellularATCommandDataSendPrefixCallback_t ) ( vo
...
+ * @ingroup cellular_common_datatypes_functionpointers
+ * @brief Undefined response callback function.
+ *
+ * @param[in] pContext The opaque cellular context pointer created by Cellular_Init.
+ * @param[in] pLine The input line form cellular modem.
+ *
+ * @return CELLULAR_PKT_STATUS_OK if the operation is successful, otherwise an error
+ * code indicating the cause of the error.
+ */
+typedef CellularPktStatus_t ( * CellularUndefinedRespCallback_t )( const CellularContext_t * pContext,
```
Comment (aggarg) [inline source/include/common/cellular_common.h:221]: Should the context be void* as in the other callback? Who provides this context?

File: test/unit-test/cellular_common_utest.c [hunk 2]
```diff
@@ -1581,3 +1592,31 @@ void test__Cellular_LibInit_PktHandlerInit_Double_Allocate_Context( void )
...
+
+/**
+ * @brief Test that happy path case for _Cellular_RegisterUndefinedRespCallback.
+ */
+void test__Cellular_RegisterUndefinedRespCallback_Happy_Path( void )
+{
+    CellularContext_t context;
+    CellularPktStatus_t pktStatus = CELLULAR_PKT_STATUS_OK;
+
+    pktStatus = _Cellular_RegisterUndefinedRespCallback( &context, cellularUndefinedRespCallback );
```
Comment (aggarg) [inline test/unit-test/cellular_common_utest.c:1620]: Is there a way to ensure that the function actually registers the callback and not just returned CELLULAR_PKT_STATUS_OK?

File: test/unit-test/cellular_pkthandler_utest.c [hunk 3]
```diff
@@ -1400,3 +1401,59 @@ void test__Cellular_AtParseInit_Check_TokenTable_Fail( void )
...
+    memset( &context, 0, sizeof( CellularContext_t ) );
+    memset( &atReq, 0, sizeof( CellularAtReq_t ) );
+    pktStatus = _Cellular_AtcmdRequestSuccessToken( &context, atReq, PACKET_REQ_TIMEOUT_MS, NULL, 1 );
+    TEST_ASSERT_EQUAL( CELLULAR_PKT_STATUS_BAD_PARAM, pktStatus );
+}
+
+/**
+ * @brief Test that _Cellular_PktioSendAtCmd return OK case for _Cellular_PktHandler_AtcmdRequestWithCallback.
+ */
+void test_test__Cellular_AtcmdRequestSuccessToken__Cellular_PktioSendAtCmd_Return_OK( void )
```
Comment (aggarg) [inline test/unit-test/cellular_pkthandler_utest.c:1438]:
```suggestion
void test__Cellular_AtcmdRequestSuccessToken_Cellular_PktioSendAtCmd_Return_OK( void )
```

File: test/unit-test/cellular_pkthandler_utest.c [hunk 4]
```diff
@@ -1400,3 +1401,59 @@ void test__Cellular_AtParseInit_Check_TokenTable_Fail( void )
...
+        sizeof( int32_t ),
+    };
+    CellularContext_t context = { 0 };
+    const char * successTokenTable[] = { "1, CONNECT CLOSE" };
+
+    _Cellular_PktioSendAtCmd_IgnoreAndReturn( CELLULAR_PKT_STATUS_OK );
+
+    /* xQueueReceive true, and the data is CELLULAR_PKT_STATUS_OK. */
+    queueData = CELLULAR_PKT_STATUS_OK;
+    pktStatus = _Cellular_AtcmdRequestSuccessToken( &context, atReqGetMccMnc, PACKET_REQ_TIMEOUT_MS, successTokenTable, 1 );
```
Comment (aggarg) [inline test/unit-test/cellular_pkthandler_utest.c:1457]: What does this test? Will an implementation that just retrurns CELLULAR_PKT_STATUS_OK pass this test?

File: test/unit-test/cellular_pktio_utest.c [hunk 5]
```diff
@@ -1192,6 +1208,68 @@ void test__Cellular_PktioInit_Thread_Rx_Data_Event_CELLULAR_AT_NO_COMMAND( void
...
+    atCmdType = CELLULAR_AT_NO_COMMAND;
+    undefineReturnStatus = CELLULAR_PKT_STATUS_OK;
+    context.undefinedRespCallback = undefinedRespCallback;
+
+    /* copy the token table. */
+    ( void ) memcpy( &context.tokenTable, &tokenTable, sizeof( CellularTokenTable_t ) );
+
+    /* Check that CELLULAR_PKT_STATUS_OK is returned. */
+    pktStatus = _Cellular_PktioInit( &context, PktioHandlePacketCallback_t );
+    TEST_ASSERT_EQUAL( CELLULAR_PKT_STATUS_OK, pktStatus );
```
Comment (aggarg) [inline test/unit-test/cellular_pktio_utest.c:1239]: Should we verify that undefinedRespCallback gets called?

File: source/include/common/cellular_common.h (hunk 1 above)
Comment (chinglee-iot) [reply source/include/common/cellular_common.h]: Update the in new commit
* _Cellular_RegisterUndefinedRespCallback accept callback context.
* CellularUndefinedRespCallback_t accept callback context.

File: test/unit-test/cellular_pkthandler_utest.c (hunk 4 above)
Comment (chinglee-iot) [reply test/unit-test/cellular_pkthandler_utest.c:1457]: This test verify _Cellular_AtcmdRequestSuccessToken returns okay if all the parameters are set correctly.
The success token table behavior is verified in test cases in cellular_pktio_utest.c.

File: test/unit-test/cellular_pktio_utest.c (hunk 5 above)
Comment (chinglee-iot) [reply test/unit-test/cellular_pktio_utest.c:1265]: The test cases is updated with the following situations
* undefinedRespCallback is called and expected string is received.
* undefinedRespCallback is called and expected string is not received.

File: test/unit-test/cellular_common_utest.c (hunk 2 above)
Comment (chinglee-iot) [reply test/unit-test/cellular_common_utest.c]: The prototype of _Cellular_RegisterUndefinedRespCallback is update with
```
CellularError_t _Cellular_RegisterUndefinedRespCallback( CellularContext_t * pContext,
                                                         CellularUndefinedRespCallback_t undefinedRespCallback,
                                                         void * pCallbackContext );
```
undefinedRespCallback and pCallbackContext can be NULL to unregister the callback and callback context.
The test case is update with the following situation
* Register the cellular undefined response callback.
* Unregister the cellular undefined response callback.
* Register the cellular undefined response callback without callback context.
* Unregister the cellular undefined response callback without callback context.
The internal filed of CellularContext is also checked in the test case to ensure the function actually registers.

**Review by ActoryOu — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**